engine/profile_tactic.ml

```diff
@@ -129,48 +129,83 @@ let format_ratio x = (Printf.sprintf "%.1f%%" (100. *. x))
 let padl n s = ws (max 0 (n - utf8_length s)) ++ str s
 let padr_with c n s =
   let ulength = utf8_length s in
-  str (utf8_sub s 0 n) ++ str (String.make (max 0 (n - ulength)) c)
+  if Int.equal n ulength then str s
+  else if n < ulength then str (utf8_sub s 0 n)
+  else str s ++ str (String.make (n - ulength) c)
 
-let rec list_iter_is_last f = function
+let rec list_map_is_last f = function
   | []      -> []
   | [x]     -> [f true x]
-  | x :: xs -> f false x :: list_iter_is_last f xs
+  | x :: xs -> f false x :: list_map_is_last f xs
 
-let header =
+let repeat_str n s =
+  if String.is_empty s then s
+  else
+    let len = String.length s in
+    String.init (n * len) (fun i -> s.[i mod len])
+
+let header name_width =
   str " tactic                                   local  total   calls       max " ++
   fnl () ++
-  str "────────────────────────────────────────┴──────┴──────┴───────┴─────────┘" ++
+  str (repeat_str name_width "─") ++ str "┴──────┴──────┴───────┴─────────┘" ++
   fnl ()
 
-let rec print_node ~filter all_total indent prefix (s, e) =
-  h (
-    padr_with '-' 40 (prefix ^ s ^ " ")
-    ++ padl 7 (format_ratio (e.local /. all_total))
-    ++ padl 7 (format_ratio (e.total /. all_total))
-    ++ padl 8 (string_of_int e.ncalls)
-    ++ padl 10 (format_sec (e.max_total))
-  ) ++
-  fnl () ++
-  print_table ~filter all_total indent false e.children
+module Line = struct
+  type t = {
+    prefix : string;
+    tac_name : string;
+    local : float;
+    total : float;
+    calls : int;
+    maxtime : float;
+  }
+
+  let pr ~name_width l =
+    h (
+      padr_with '-' name_width (l.prefix ^ l.tac_name ^ " ")
+      ++ padl 7 (format_ratio l.local)
+      ++ padl 7 (format_ratio l.total)
+      ++ padl 8 (string_of_int l.calls)
+      ++ padl 10 (format_sec l.maxtime))
+end
 
-and print_table ~filter all_total indent first_level table =
+let rec linearize_node ~filter all_total indent prefix (s, e) =
+  { Line.prefix; tac_name=s;
+    local = (e.local /. all_total);
+    total = (e.total /. all_total);
+    calls = e.ncalls;
+    maxtime = e.max_total;
+  } :: linearize_table ~filter all_total indent false e.children
+
+and linearize_table ~filter all_total indent first_level table =
   let fold _ n l =
     let s, total = n.name, n.total in
     if filter s total then (s, n) :: l else l in
   let ls = M.fold fold table [] in
   match ls with
   | [s, n] when not first_level ->
-     v 0 (print_node ~filter all_total indent (indent ^ "└") (s, n))
+    linearize_node ~filter all_total indent (indent ^ "└") (s, n)
```

Comment:
Was dropping the `v 0` here and not adding it anywhere else deliberate?

Reply:
yes because I can't tell what it's supposed to be for.